[Don't Merge] Deprecated: Restructure docs for website

jlin27 opened the pull request:
***Note:*** This is not the latest version of docs changes. Please see PR: https://github.com/pytorch/serve/pull/624

---

This is one of three PRs that will setup the automated docs build:

* This PR modifies the content (.rst, .md) files
* PR #274: Add sphinx directory and conf.py
* PR #275: Add .github/actions with npm modules

---
Per the [suggestions here](https://github.com/pytorch/serve/pull/254#pullrequestreview-399373186), this PR:
* Moves contents from docs/README.md into docs/index.rst (keeping as .rst to leverage the rst warning label) and changes the relative links to absolute urls
* Renames  ~/README.md to install.md
* Moves content from model-archiver/README.md into model-archiver.md and updates relative links to absolute urls 


Here is a screenshot of the new homepage: 

![image](https://user-images.githubusercontent.com/8042156/80772170-a9a02500-8b0a-11ea-8c87-a937e359f469.png)

aaronmarkham commented on docs/model-archiver.md:
```diff
@@ -37,7 +37,7 @@ pip install .
 Now let's cover the details on using the CLI tool: `model-archiver`.
 
-Here is an example usage with the squeezenet_v1.1 model archive following the example in the [examples README](../examples/README.md):
+Here is an example usage with the squeezenet_v1.1 model archive following the example in the [examples README](https://github.com/pytorch/serve/tree/master/examplesZ):
```
```suggestion
Here is an example usage with the squeezenet_v1.1 model archive following the example in the [examples README](https://github.com/pytorch/serve/tree/master/examples):
```

jlin27 replied:
Cleaned up syntax

aaronmarkham commented on docs/index.rst:
```diff
@@ -0,0 +1,28 @@
+Basic Features
+----------------
+
+* `Serving Quick Start <https://pytorch.org/serve/install.html#serve-a-model>`_ - Basic server usage tutorial
```
I'm thinking all of these can be relative links.

jlin27 replied:
Updated to relative link

aaronmarkham commented on docs/index.rst:
```diff
@@ -0,0 +1,28 @@
+* `Serving Quick Start <https://pytorch.org/serve/install.html#serve-a-model>`_ - Basic server usage tutorial
+* `Model Archive Quick Start <https://pytorch.org/serve/model-archiver.html#creating-a-model-archive>`_ - Tutorial that shows you how to package a model archive file.
+* `Installation <https://pytorch.org/serve/install.html##install-torchserve>`_ - Installation procedures
+* `Serving Models <https://pytorch.org/serve/server.md>`_ - Explains how to use `torchserve`.
```
All of these could probably be .html instead of .md

jlin27 replied:
Updated to relative link that references the .md file

aaronmarkham commented on docs/index.rst:
```diff
@@ -0,0 +1,28 @@
+
+* `Advanced settings <https://pytorch.org/serve/configuration.md>`_ - Describes advanced TorchServe configurations.
+* `Custom Model Service <https://pytorch.org/serve/custom_service.md>`_ - Describes how to develop custom inference services.
+* `Unit Tests <https://github.com/pytorch/serve/tree/master/ts/tests/README.md>`_ - Housekeeping unit tests for TorchServe.
```
Stuff outside of /docs would need to be statically linked and to .md since they won't get transpiled by Sphinx.
(So this is fine as is... unlike the previous links)

jlin27 replied:
Kept the README.md files as links

aaronmarkham commented on docs/install.md:
````diff
@@ -0,0 +1,262 @@
+1. Create an environment and install torchserve and torch-model-archiver
+    For CPU
+    ```bash
+    conda create --name torchserve torchserve torch-model-archiver pytorch torchtext torchvision -c pytorch -c powerai
````
Will need to grab the latest from #286.

jlin27 replied:
Pulled in latest install changes from PR#286

jlin27 commented:
@aaronmarkham  - I addressed the suggestions you made. I only have a local preview. To built it locally, I'd suggest grabbing the changes from this PR as well as the changes from PR https://github.com/pytorch/serve/pull/274 which contains the sphinx + conf.py.

aaronmarkham commented on docs/model-archiver.md:
````diff
@@ -159,4 +159,4 @@ For eagermode:
 ```
 
-This will package all the model artifacts files and output `densenet_161.mar` in the current working directory. This `.mar` file is all you need to run TorchServe, serving inference requests for a simple image recognition API. Go back to the [Serve a Model tutorial](../README.md#serve-a-model) and try to run this model archive that you just created!
+This will package all the model artifacts files and output `densenet_161.mar` in the current working directory. This `.mar` file is all you need to run TorchServe, serving inference requests for a simple image recognition API. Go back to the [Serve a Model tutorial](https://github.com/pytorch/serve/blob/master/README.md#serve-a-model) and try to run this model archive that you just created!
````
This should link to the same thing you have on index.rst: 
```
:ref:`Serving Quick Start <install:Serve a model>` - Basic server usage tutorial
```

jlin27 replied:
I think that syntax only works for .rst files. I'm not aware of a way to link to s subheader in another file in a .md file. 
The alternatives I would suggest are:
* `[Serve a Model tutorial](install.md)`
* `[Serve a Model tutorial](https://pytorch.org/serve/docs/sphinx/build/html/install.html)`

aaronmarkham commented on docs/index.rst:
```diff
@@ -0,0 +1,29 @@
+Basic Features
+----------------
+
+* :ref:`Serving Quick Start <install:Serve a model>` - Basic server usage tutorial
```
Do we have to use `ref`? I'm not that familiar with this annotation, but it pretty much makes the index.rst file in the repo unusable from GitHub. Seems like it'll only work when rendered by Sphinx...

aaronmarkham replied:
Not a show stopper if we're going towards the docs only being used from the website. But if so, the tagline for the project should point to the web version of the docs. (and we probably should keep a README.md in the docs folder with the info on how the docs are built and a link to the most recent output of the docs.)

jlin27 commented on docs/model-archiver.md:
```diff
@@ -117,7 +117,7 @@ optional arguments:
 A valid model name must begin with a letter of the alphabet and can only contains letters, digits, underscores `_`, dashes `-` and periods `.`.
 
-**Note**: The model name can be overridden when you register the model with [Register Model API](../docs/management_api.md#register-a-model).
+**Note**: The model name can be overridden when you register the model with [Register Model API](https://pytorch.org/serve/management_api.html#register-a-models).
```
Were you able to get this to work when running locally?
When I tried, it incorrectly resolves to `https://pytorch.org/serve/management_api.md#register-a-models` (has a .md instead of .html). I think it's because sphinx for .md doesn't support linking to a subheader in another file.

jlin27 replied:
I wasn't able to get that new syntax `[Register Model API](management_api.md#register-a-models)` to resolve correctly - this new syntax resolves to resolves `https://pytorch.org/serve/management_api.mdl#register-a-models` instead of a .html file.
Were you able to get this to work when you built it?
What I'd used in the index.rst was the [sphinx autosectionlabel directive](https://www.sphinx-doc.org/en/master/usage/extensions/autosectionlabel.html) but I think it only applies to .rst files.

aaronmarkham commented:
Added comments in #274 for some Sphinx conf and requirements.txt changes... there's overlap in the PRs so you'd want to align them or maybe keep the files separate between each PR.

jlin27 commented:
@aaronmarkham - Thanks for the review. I removed conf.py from this PR and addressed the sphinx and recommonmark versions in the other PR: https://github.com/pytorch/serve/pull/274

Are we set for me to merge this once the tests pass?

chauhang commented on docs/model-archiver.md:
```diff
@@ -37,7 +37,7 @@ pip install .
 Now let's cover the details on using the CLI tool: `model-archiver`.
 
-Here is an example usage with the squeezenet_v1.1 model archive following the example in the [examples README](../examples/README.md):
+Here is an example usage with the squeezenet_v1.1 model archive following the example in the [examples README](https://github.com/pytorch/serve/tree/master/examples):
```
@jlin27 Can we keep the links relative and not hardcode the path?

jlin27 replied:
Hi Geeta - This link will point to the examples section in the repo, rather than a set of documentation so that is why it is hardcoded. The reason it was a relative link in the original was because it was still referencing within the repo.

chauhang commented on docs/model-archiver.md:
```diff
@@ -135,7 +135,7 @@ Handler can be TorchServe's inbuilt handler name or path to a py to handle custo
 4. `image_segmenter`
 
-For more details refer [default handler documentation](../docs/default_handlers.md)
+For more details refer [default handler documentation](https://pytorch.org/serve/default_handlers.html)
```
@jlin27 Same comment here to keep all the links relative and not hardcode the path

jlin27 replied:
This was kept as a hardcoded link because we found that in md files for sphinx, relative links to other document did not resolve correctly.

chauhang commented on docs/model-archiver.md:
````diff
@@ -159,4 +159,4 @@ For eagermode:
 ```
 
-This will package all the model artifacts files and output `densenet_161.mar` in the current working directory. This `.mar` file is all you need to run TorchServe, serving inference requests for a simple image recognition API. Go back to the [Serve a Model tutorial](../README.md#serve-a-model) and try to run this model archive that you just created!
+This will package all the model artifacts files and output `densenet_161.mar` in the current working directory. This `.mar` file is all you need to run TorchServe, serving inference requests for a simple image recognition API. Go back to the [Serve a Model tutorial](https://github.com/pytorch/serve/blob/master/README.md#serve-a-model) and try to run this model archive that you just created!
````
@jlin27 Can we keep all the links as relative?

jlin27 replied:
Hi Geeta - These were intentionally kept as absolute links because they point to the github repo, rather than documentation.

msaroufim commented:
Hi @jlin27 are you still working on this? There’s a couple of open issues around broken links. Let me know if we should assign this to someone else.
